tests/test_trial.py

```diff
@@ -764,9 +765,329 @@ def test_study_id():
     assert trial._study_id == trial.study._study_id
 
 
-def test_frozen_trial_validate():
+def test_frozen_trial_eq_ne():
+    # type: () -> None
+
+    trial = _create_frozen_trial()
+
+    trial_other = copy.copy(trial)
+    assert trial == trial_other
+
+    trial_other.value = 0.3
+    assert trial != trial_other
+
+
+def test_frozen_trial_lt():
+    # type: () -> None
+
+    trial = _create_frozen_trial()
+
+    trial_other = copy.copy(trial)
+    assert not trial < trial_other
+
+    trial_other.number = trial.number + 1
+    assert trial < trial_other
+    assert not trial_other < trial
+
+    with pytest.raises(TypeError):
+        trial < 1
+
+    assert trial <= trial_other
+    assert not trial_other <= trial
+
+    with pytest.raises(TypeError):
+        trial <= 1
+
+    # A list of FrozenTrials is sortable.
+    trials = [trial_other, trial]
+    trials.sort()
+    assert trials[0] is trial
+    assert trials[1] is trial_other
+
+
+def _create_frozen_trial():
+    # type: () -> FrozenTrial
+
+    return FrozenTrial(
+        number=0,
+        trial_id=0,
+        state=TrialState.COMPLETE,
+        value=0.2,
+        datetime_start=datetime.datetime.now(),
+        datetime_complete=datetime.datetime.now(),
+        params={"x": 10},
+        distributions={"x": UniformDistribution(5, 12)},
+        user_attrs={},
+        system_attrs={},
+        intermediate_values={},
+    )
+
+
+def test_frozen_trial_repr():
     # type: () -> None
 
+    trial = FrozenTrial(
+        number=0,
+        trial_id=0,
+        state=TrialState.COMPLETE,
+        value=0.2,
+        datetime_start=datetime.datetime.now(),
+        datetime_complete=datetime.datetime.now(),
+        params={"x": 10},
+        distributions={"x": UniformDistribution(5, 12)},
+        user_attrs={},
+        system_attrs={},
+        intermediate_values={},
+    )
+
+    assert trial == eval(repr(trial))
+
+
+@parametrize_storage
+def test_frozen_trial_sampling(storage_init_func):
+    # type: (Callable[[], storages.BaseStorage]) -> None
+
+    def objective(trial: BaseTrial) -> float:
```

Comment:
Oh, the mypy type checking passed with `BaseTrial`? I thought we had to change it given [`Study.optimize`](https://github.com/optuna/optuna/blob/master/optuna/study.py#L213).

Reply:
Yes, it is...

I'm not sure, but how about this?
```python
    def objective(trial):
        # type: (Union[Trial, FrozenTrial]) -> float
```

Reply:
I think we should relax [`ObjectiveFuncType`](https://github.com/optuna/optuna/blob/master/optuna/study.py#L42) to accept `BaseTrial`s and not only `Trial`s since the public APIs for both types are identical (and are likely not going to diverge in a non-compatible way the near future).

Was mypy not complaining on the current (in this PR) test? If so, we could do that in a follow-up PR. Open for feedback.

Reply:
Thank you for your clarification! Yes, mypy was passed in this PR: 
https://app.circleci.com/pipelines/github/optuna/optuna/1711/workflows/96a2a01e-6ee5-4812-a31d-5ef182218b57/jobs/49893/parallel-runs/0/steps/0-106
I agree with you.

Reply:
Created an issue so that we don't forget, https://github.com/optuna/optuna/issues/1577.